src/builder.rs

```diff
@@ -0,0 +1,213 @@
+    ///
+    /// let email = redactor.redact(email);
+    /// let name = redactor.redact(name);
+    ///
+    /// assert_eq!(
+    ///     format!("{} <{}>", name, email),
+    ///     "JoXX Xoe <johX.XXX@XXXXa.it>"
+    /// );
+    /// ```
+    pub fn redact(&self, data: String) -> String {
+        give_me_a_formatter!(
+            move<'a> {
+                data: &'a dyn Display = &data.as_str(),
+                flags: RedactFlags = self.0
+            }
+
+            fn fmt(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                std::fmt::Debug::fmt(
+                    &RedactionFormatter {
+                        this: RedactionTarget::Display(data),
+                        flags: *flags,
+                        specialization: None,
+                    },
+                    fmt,
+                )
+            }
+        )
+    }
+
+    /// Redact the given string in-place.
+    ///
+    /// Convenience method for chaining.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use veil::RedactorBuilder;
+    /// let mut email = "john.doe@prima.it".to_string();
+    /// let mut name = "John Doe".to_string();
+    ///
+    /// RedactorBuilder::new()
+    ///     .char('X')
+    ///     .partial()
+    ///     .build()
+    ///     .unwrap()
+    ///     .and_redact(&mut email)
+    ///     .and_redact(&mut name);
+    ///
+    /// assert_eq!(
+    ///     format!("{} <{}>", name, email),
+    ///     "JoXX Xoe <johX.XXX@XXXXa.it>"
+    /// );
+    /// ```
+    pub fn and_redact(&self, data: &mut String) -> &Self {
+        *data = self.redact(core::mem::take(data));
+        self
+    }
+
+    /// Wrap the given data in a [`RedactWrapped`], allowing it to be redacted when displayed or debugged.
+    ///
+    /// Currently, the only supported [`Debug`] formats are `{:?}` and `{:#?}`. Other flags will be ignored.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # use veil::RedactorBuilder;
+    /// let email = "john.doe@prima.it".to_string();
+    /// let name = "John Doe".to_string();
+    ///
+    /// let redactor = RedactorBuilder::new()
+    ///     .char('X')
+    ///     .partial()
+    ///     .build()
+    ///     .unwrap();
+    ///
+    /// let email = redactor.wrap(&email);
+    /// let name = redactor.wrap(&name);
+    ///
+    /// assert_eq!(
+    ///     format!("{} <{}>", name, email),
+    ///     "JoXX Xoe <johX.XXX@XXXXa.it>"
+    /// );
+    ///
+    /// assert_eq!(
+    ///     format!("{:?} <{:#?}>", name, email),
+    ///     "\"JoXX Xoe\" <\"johX.XXX@XXXXa.it\">"
+    /// );
+    /// ```
+    pub const fn wrap<'a, T>(&'a self, data: &'a T) -> RedactWrapped<'a, T> {
+        RedactWrapped { flags: &self.0, data }
+    }
+}
+
+/// A checked builder for [`Redactor`]s.
+pub struct RedactorBuilder {
+    redact_char: Option<char>,
+    partial: bool,
+}
+impl RedactorBuilder {
+    /// Initialize a new redaction flag builder.
+    #[inline(always)]
```

Comment:
No no, I checked this one with Godbolt. It doesn't optimise away the compile time checks without the inlines.

Reply:
that doesn't seem to be the case even with opt_level 1.
https://godbolt.org/z/5WqMaqvav

I don't think we care about inlining here in dev, since it might slow down compilation

Reply:
https://godbolt.org/z/nfxcK7EKs

If you take the `#[inline(always)]` away from `build`, you can see that it just fails to const evaluate for some reason

Reply:
I have an absurd level of faith in compilers, so my heart is telling me we don't need to do this, but I can't argue with the facts.

Reply:
This particular case also caused me psychological damage at the time

I'm here for you if you need support